<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GWAS.utils • GWAS.utils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GWAS.utils">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GWAS.utils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/gwas-utils-package.html">GWAS.utils</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sinarueeger/GWAS.utils">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>GWAS.utils</h1>
            <h3 class="subtitle">Quick introduction to GWAS.utils</h3>
                        <h4 class="author">Sina Rueger</h4>
            
            <h4 class="date">2019-07-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sinarueeger/GWAS.utils/blob/master/vignettes/gwas-utils-package.Rmd"><code>vignettes/gwas-utils-package.Rmd</code></a></small>
      <div class="hidden name"><code>gwas-utils-package.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(GWAS.utils)</span></code></pre></div>
<p><code>GWAS.utils</code> is an R package with basic helper functions for manipulating GWAS data, including two GWAS datasets.</p>
<p>The functions are basic, some just trivial wrappers with a few lines of code (e.g. <code><a href="../reference/trans_inv_normal.html">trans_inv_normal()</a></code>). But they all tackle a specific problem, for which either no function was elsewhere available, or the function was available, but part of a heavy package.</p>
<p>Here is a quick overview of what functions are available.</p>
<ul>
<li>
<code><a href="../reference/eaf2maf.html">eaf2maf(eaf = eaf_vec)</a></code>: Transforms effect allele frequencies into minor allele frequencies.</li>
<li>
<code>inv_normal(x = vec)</code>: Inverse normal transformation of a variable.</li>
<li>
<code><a href="../reference/z2p.html">z2p(z = z_vec)</a></code>: Transforms Z-statistics to P-values, includes an option for very large Z-statistics.</li>
<li>
<code><a href="../reference/eff_nbr_tests.html">eff_nbr_tests(mat = genotype_df)</a></code>: Calculates the effective number of tests of a GWAS, based on the correlation between the SNPs.</li>
<li>
<code><a href="../reference/QQplot.html">QQplot(p = p_vec)</a></code>: Q-Q-plot of P-values (uniformly distributed under the null).</li>
<li>
<code><a href="../reference/genomic_inflation.html">genomic_inflation(Z = z_vec)</a></code>: Calculates genomic inflation, with either P-values or Z-statistics.</li>
</ul>
<p>Further down you will find examples for each of them.</p>
<p>The usefulness of this package for other users, however, are two datasets. Before going into the details of the functions, let me first describe the datasets.</p>
<div id="data" class="section level2">
<h2 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h2>
<p>To my knowledge, there are no real (i.e. not simulated) genetic datasets available in R, hence there are also just a few R examples around that use genetic data. There are two principal reasons for the lack of genetic data in R. First, genetic data is usually securely locked and not publicly available (for good reasons!). Second, genetic data is typically large in size, often hundreds of thousands of columns wide, making it suboptimal to embedded in an R package. The two datasets contained in the <code>GWAS.utils</code> package are open source, and restricted in size.</p>
<div id="opensnp-genotype-data" class="section level3">
<h3 class="hasAnchor">
<a href="#opensnp-genotype-data" class="anchor"></a>openSNP genotype data</h3>
<p>There is a public genetic database called <a href="https://opensnp.org/">openSNP</a>, and for the purpose of the <a href="https://www.crowdai.org/">crowdAI</a> <em>height prediction challenge</em> <a href="https://github.com/onaret">Olivier Naret</a> curated a genotype dataset that is available for download on <a href="https://zenodo.org/record/1442755#.XLCKkKaxXUI">zenodo</a> (Naret 2018).</p>
<p>To generate the dataset <code>opensnp</code>, I downloaded the dataset from <a href="https://zenodo.org/record/1442755#.XLCKkKaxXUI">zenodo</a>, extracted 21 SNPs on chromosome 1 within the range of 15’000’000 and 20’000’000 bp<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> from the training dataset of 784 individuals and transformed the vcf file with <a href="http://zzz.bwh.harvard.edu/plink/">PLINK</a> (Chang <em>et al.</em> 2015) into the number of minor alleles (allele dosages): 0 for homozygous major allele (AA), 1 heterozygous (Aa), 2 homozygous minor allele (aa). I then joined the dataset with the height information that was also available and stored the dataset as <code>opensnp</code>.</p>
<p>The data preparation process is described in <a href="https://github.com/sinarueeger/GWAS.utils/blob/master/data-raw/opensnp-data.R">opensnp-data.R</a>.</p>
<p>Each row represents one individual, the columns the ID, height and 22 genetic variants (SNPs). The SNP name is a combination of SNP identifier and the minor allele.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>skimr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/skimr/topics/skim_with">skim_with</a></span>(<span class="dt">factor =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">ordered =</span> <span class="ot">NULL</span>))</span>
<span id="cb2-2"><a href="#cb2-2"></a>skimr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/skimr/topics/skim">skim</a></span>(opensnp <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="op">-</span>id) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">mutate_at</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/vars.html">vars</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">starts_with</a></span>(<span class="st">"rs"</span>)), as.factor))</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co">#&gt; Skim summary statistics</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co">#&gt;  n obs: 784 </span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="co">#&gt;  n variables: 22 </span></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co">#&gt; </span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co">#&gt; ── Variable type:factor ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co">#&gt;      variable missing complete   n n_unique                    top_counts</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co">#&gt;  rs10489858_C       0      784 784        3   0: 627, 1: 154, 2: 3, NA: 0</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co">#&gt;  rs10917451_C       0      784 784        3  0: 400, 1: 321, 2: 63, NA: 0</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">#&gt;  rs12059827_G       0      784 784        3    0: 691, 1: 91, 2: 2, NA: 0</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co">#&gt;  rs12066114_G       0      784 784        3  0: 411, 1: 317, 2: 56, NA: 0</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="co">#&gt;   rs1542875_C       0      784 784        3  0: 384, 1: 329, 2: 71, NA: 0</span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="co">#&gt;   rs1985272_C       0      784 784        3 1: 375, 0: 276, 2: 133, NA: 0</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="co">#&gt;   rs2014725_T       0      784 784        3 1: 349, 0: 303, 2: 132, NA: 0</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="co">#&gt;   rs2057096_G       0      784 784        3 1: 372, 0: 266, 2: 146, NA: 0</span></span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="co">#&gt;   rs2076595_T       0      784 784        3 1: 378, 0: 278, 2: 128, NA: 0</span></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="co">#&gt;   rs2076596_T       0      784 784        3 1: 377, 0: 279, 2: 128, NA: 0</span></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="co">#&gt;    rs214302_G       0      784 784        3 1: 396, 0: 238, 2: 150, NA: 0</span></span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="co">#&gt;   rs3748756_T       0      784 784        3  0: 474, 1: 266, 2: 44, NA: 0</span></span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="co">#&gt;   rs3818033_G       0      784 784        3 0: 341, 1: 341, 2: 102, NA: 0</span></span>
<span id="cb2-22"><a href="#cb2-22"></a><span class="co">#&gt;   rs3818038_G       0      784 784        3 0: 342, 1: 340, 2: 102, NA: 0</span></span>
<span id="cb2-23"><a href="#cb2-23"></a><span class="co">#&gt;   rs4508039_C       0      784 784        3 1: 373, 0: 306, 2: 105, NA: 0</span></span>
<span id="cb2-24"><a href="#cb2-24"></a><span class="co">#&gt;   rs4911997_A       0      784 784        3 1: 390, 0: 238, 2: 156, NA: 0</span></span>
<span id="cb2-25"><a href="#cb2-25"></a><span class="co">#&gt;   rs4912090_C       0      784 784        3  0: 461, 1: 271, 2: 52, NA: 0</span></span>
<span id="cb2-26"><a href="#cb2-26"></a><span class="co">#&gt;    rs730387_A       0      784 784        3  0: 380, 1: 319, 2: 85, NA: 0</span></span>
<span id="cb2-27"><a href="#cb2-27"></a><span class="co">#&gt;   rs7516087_G       0      784 784        3  0: 396, 1: 310, 2: 78, NA: 0</span></span>
<span id="cb2-28"><a href="#cb2-28"></a><span class="co">#&gt;   rs7536679_G       0      784 784        3 1: 386, 0: 218, 2: 180, NA: 0</span></span>
<span id="cb2-29"><a href="#cb2-29"></a><span class="co">#&gt;    rs999974_G       0      784 784        3  0: 385, 1: 331, 2: 68, NA: 0</span></span>
<span id="cb2-30"><a href="#cb2-30"></a><span class="co">#&gt; </span></span>
<span id="cb2-31"><a href="#cb2-31"></a><span class="co">#&gt; ── Variable type:numeric ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span>
<span id="cb2-32"><a href="#cb2-32"></a><span class="co">#&gt;  variable missing complete   n   mean   sd     p0 p25 p50 p75  p100</span></span>
<span id="cb2-33"><a href="#cb2-33"></a><span class="co">#&gt;    height       0      784 784 173.21 10.5 142.24 165 173 181 203.2</span></span>
<span id="cb2-34"><a href="#cb2-34"></a><span class="co">#&gt;      hist</span></span>
<span id="cb2-35"><a href="#cb2-35"></a><span class="co">#&gt;  ▁▂▇▆▇▆▂▁</span></span></code></pre></div>
</div>
<div id="human-height-summary-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#human-height-summary-statistics" class="anchor"></a>Human height summary statistics</h3>
<p>The second dataset are summary statistics from a height meta-analysis published in 2018 by <a href="https://academic.oup.com/hmg/article/27/20/3641/5067845">Yengo <em>et al. 2018</em></a> (download <a href="https://portals.broadinstitute.org/collaboration/giant/images/c/c8/Meta-analysis_Locke_et_al%2BUKBiobank_2018_UPDATED.txt.gz?download=1">here</a>), estimated from over 600’000 individuals. The 10’000 genetic variants are a random sample from 2’336’269 genetic variants in the dataset.</p>
<p>Each row represents one genetic variant (SNP), the columns the chromosome (<code>CHR</code>), position in bp (<code>POS</code>), the SNP identifier (<code>SNP</code>), the effect allele (<code>Tested_Allele</code>), the reference allele (<code>Other_Allele</code>), effect allele frequency in the Health and Retirement Study (HRS) (<code>Freq_Tested_Allele_in_HRS</code>), effect size (<code>BETA</code>), standard error (<code>SE</code>), p-value (<code>P</code>) and sample size (<code>N</code>).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(giant)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="co">#&gt; Classes 'tbl_df', 'tbl' and 'data.frame':    10000 obs. of  10 variables:</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">#&gt;  $ CHR                      : int  4 14 3 2 13 20 10 4 11 14 ...</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt;  $ POS                      : int  54901476 35406913 107654661 196176509 22008438 44381615 127920542 42646677 11986061 52176988 ...</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt;  $ SNP                      : chr  "rs11934081" "rs712313" "rs17228803" "rs1554603" ...</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt;  $ Tested_Allele            : chr  "T" "T" "A" "T" ...</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt;  $ Other_Allele             : chr  "C" "G" "G" "G" ...</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt;  $ Freq_Tested_Allele_in_HRS: num  0.637 0.71 0.384 0.965 0.68 ...</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt;  $ BETA                     : num  0.0004 0.0008 -0.0029 -0.0056 0.0044 -0.0008 0.0009 0.0026 0.0101 -0.0038 ...</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">#&gt;  $ SE                       : num  0.0018 0.0019 0.0018 0.0048 0.0019 0.0018 0.0018 0.002 0.002 0.0021 ...</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#&gt;  $ P                        : num  8.0e-01 6.8e-01 9.5e-02 2.5e-01 1.8e-02 6.5e-01 6.2e-01 1.9e-01 3.7e-07 6.8e-02 ...</span></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co">#&gt;  $ N                        : int  691061 686354 691299 635193 690898 692535 686996 672011 692536 691290 ...</span></span></code></pre></div>
</div>
</div>
<div id="functions" class="section level2">
<h2 class="hasAnchor">
<a href="#functions" class="anchor"></a>Functions</h2>
<p>The six functions can be categorised into two groups: four of them take some type of summary statistics as input, two of them individual data (<code>inv_normal()</code> and <code>eff_nbr_tests</code>).</p>
<div id="effective-number-of-tests-performed" class="section level3">
<h3 class="hasAnchor">
<a href="#effective-number-of-tests-performed" class="anchor"></a>Effective number of tests performed</h3>
<p>To account for multiple testing, Bonferroni correction is often used in the context of GWASs. But because of correlation between genetic variants (=<a href="https://en.wikipedia.org/wiki/Linkage_disequilibrium">linkage disequilibrium</a>), the number of tests performed might over-adjust the significance threshold. Using a method by <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3055936/">Gao <em>et al.</em> 2008</a> we can determine the <em>effective</em> number of tests performed, based on the correlation between the SNPs. The input of the function is a genotype dataset or a correlation matrix, the output a single number.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw"><a href="../reference/eff_nbr_tests.html">eff_nbr_tests</a></span>(<span class="dt">mat =</span> opensnp <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="op">-</span>id, <span class="op">-</span>height))</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co">#&gt; [1] 16</span></span></code></pre></div>
</div>
<div id="inverse-normal-transformation" class="section level3">
<h3 class="hasAnchor">
<a href="#inverse-normal-transformation" class="anchor"></a>Inverse normal transformation</h3>
<p>This function forces any variable to be normally distributed.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>opensnp<span class="op">$</span>height_tranformed &lt;-<span class="st"> </span>GWAS.utils<span class="op">::</span><span class="kw">trans_inv_normal</span>(opensnp<span class="op">$</span>height)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/par">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/qqnorm">qqnorm</a></span>(opensnp<span class="op">$</span>height, <span class="dt">main =</span> <span class="st">"Untransformed height variable"</span>)</span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/qqnorm">qqline</a></span>(opensnp<span class="op">$</span>height)</span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/qqnorm">qqnorm</a></span>(opensnp<span class="op">$</span>height_tranformed, <span class="dt">main =</span> <span class="st">"Untransformed height variable"</span>)</span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/qqnorm">qqline</a></span>(opensnp<span class="op">$</span>height_tranformed)</span></code></pre></div>
<p><img src="gwas-utils-package_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="co">## or</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(<span class="dv">1000</span>)</span>
<span id="cb6-4"><a href="#cb6-4"></a>X_transformed &lt;-<span class="st"> </span>GWAS.utils<span class="op">::</span><span class="kw">trans_inv_normal</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(<span class="dv">1000</span>))</span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/par">par</a></span>(<span class="dt">mfrow =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/qqnorm">qqnorm</a></span>(X, <span class="dt">main =</span> <span class="st">"Uniform distributed variable"</span>)</span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/qqnorm">qqline</a></span>(X)</span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/qqnorm">qqnorm</a></span>(X_transformed, <span class="dt">main =</span> <span class="st">"Transformed, formerly uniform distributed variable"</span>)</span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/qqnorm">qqline</a></span>(X_transformed)</span></code></pre></div>
<p><img src="gwas-utils-package_files/figure-html/unnamed-chunk-5-2.png" width="768"></p>
</div>
<div id="transformation-of-effect-to-minor-allele-frequency" class="section level3">
<h3 class="hasAnchor">
<a href="#transformation-of-effect-to-minor-allele-frequency" class="anchor"></a>Transformation of effect to minor allele frequency</h3>
<p>Simple function to convert effect allele frequencies into minor allele frequencies that range between 0 and 0.5</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>giant<span class="op">$</span>maf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/eaf2maf.html">eaf2maf</a></span>(giant<span class="op">$</span>Freq_Tested_Allele_in_HRS)</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(giant<span class="op">$</span>Freq_Tested_Allele_in_HRS, giant<span class="op">$</span>maf)</span></code></pre></div>
<p><img src="gwas-utils-package_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="transformation-from-z-statistics-to-p-value" class="section level3">
<h3 class="hasAnchor">
<a href="#transformation-from-z-statistics-to-p-value" class="anchor"></a>Transformation from Z-statistics to P-value</h3>
<p>In the simplest case, transforming a Z-statistic into a P-value is straightforward.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="dv">2</span> <span class="op">*</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">pnorm</a></span>(<span class="dv">2</span>, <span class="dt">lower.tail=</span> <span class="ot">FALSE</span>)</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co">#&gt; [1] 0.04550026</span></span></code></pre></div>
<p>But in genetics, P-values can get extremely small, and starting from an absolute Z-statistics of 38, the returned P-value will be 0.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">pnorm</a></span>(<span class="dv">37</span>, <span class="dt">lower.tail=</span> <span class="ot">FALSE</span>)</span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="co">#&gt; [1] 5.725571e-300</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">pnorm</a></span>(<span class="dv">38</span>, <span class="dt">lower.tail=</span> <span class="ot">FALSE</span>)</span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>There is an additional trick, using the log argument and then back transforming it using <code>exp</code>, to compute the P-value for Z=38, but for Z=39 this does not work anymore.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">exp</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">pnorm</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(<span class="dv">38</span>), <span class="dt">log.p =</span> <span class="ot">TRUE</span>, <span class="dt">lower =</span> <span class="ot">FALSE</span>)) <span class="op">*</span><span class="st"> </span><span class="dv">2</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="co">#&gt; [1] 5.770857e-316</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">exp</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">pnorm</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(<span class="dv">39</span>), <span class="dt">log.p =</span> <span class="ot">TRUE</span>, <span class="dt">lower =</span> <span class="ot">FALSE</span>)) <span class="op">*</span><span class="st"> </span><span class="dv">2</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>This has to do with the way R can operate small numbers. The package <code>Rmpfr</code> provides a solution to this, and the function <code>z2p</code> makes use of this functionality in order to deal with very small number.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw"><a href="../reference/z2p.html">z2p</a></span>(<span class="dv">39</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="co">#&gt; Some P-values are equal to 0. Try using the option method = 'Rmpfr::pnorm'</span></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="co">#&gt; [1] 0</span></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="kw"><a href="../reference/z2p.html">z2p</a></span>(<span class="dv">39</span>,  <span class="dt">method =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Rmpfr::pnorm"</span>))</span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co">#&gt; using method Rmpfr::</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="co">#&gt; 1 'mpfr' number of precision  100   bits </span></span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="co">#&gt; [1] 1.070623822430189070363097933135e-332</span></span></code></pre></div>
<p>Even though the Z-statistics in the <code>giant</code> data set are not that large, we can apply the <code>z2p</code> function to compare the known P-value with the calculated P-value.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>giant<span class="op">$</span>P_z2p_function &lt;-<span class="st"> </span><span class="kw"><a href="../reference/z2p.html">z2p</a></span>(giant<span class="op">$</span>BETA<span class="op">/</span>giant<span class="op">$</span>SE, <span class="dt">method =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Rmpfr::pnorm"</span>))</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="co">#&gt; using method Rmpfr::</span></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(<span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(giant<span class="op">$</span>P), <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(giant<span class="op">$</span>P_z2p_function), <span class="dt">main =</span> <span class="st">"Comparison of kown P-value with the</span><span class="ch">\n</span><span class="st">calculated P-value through z2p()"</span>, <span class="dt">ylab =</span> <span class="st">"P_z2p_function"</span>, <span class="dt">xlab =</span> <span class="st">"P"</span>)</span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/abline">abline</a></span>(<span class="dt">a =</span> <span class="dv">0</span>, <span class="dt">b =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="gwas-utils-package_files/figure-html/unnamed-chunk-11-1.png" width="384"></p>
</div>
<div id="qqplot-of-p-values" class="section level3">
<h3 class="hasAnchor">
<a href="#qqplot-of-p-values" class="anchor"></a>QQplot of P-values</h3>
<p>P-values should be uniformly distributed under the null hypothesis. A Q-Q-plot checks the deviation from the uniform distribution. There are a number of packages that implement this type of plot (e.g. <a href="https://cran.r-project.org/web/packages/qqman/index.html">qqman</a>).</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw"><a href="../reference/QQplot.html">QQplot</a></span>(giant<span class="op">$</span>P, <span class="dt">main =</span> <span class="st">"Random sample of 10K SNPs from</span><span class="ch">\n</span><span class="st">GIANT height summary statistics"</span>)</span></code></pre></div>
<p><img src="gwas-utils-package_files/figure-html/unnamed-chunk-12-1.png" width="384"></p>
</div>
<div id="genomic-inflation-factor" class="section level3">
<h3 class="hasAnchor">
<a href="#genomic-inflation-factor" class="anchor"></a>Genomic inflation factor</h3>
<p>To check a GWAS for inflation in P-values, the <a href="https://en.wikipedia.org/wiki/Population_stratification">genomic inflation factor</a> can be computed, which should be around a value of 1. Compared to other genomic inflation implementations, our function takes two types of summary statistics as input by making an assumption about the P-value origin.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co">## use with Z-statistics</span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="kw"><a href="../reference/genomic_inflation.html">genomic_inflation</a></span>(<span class="dt">Z =</span> giant<span class="op">$</span>BETA<span class="op">/</span>giant<span class="op">$</span>SE)</span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="co">#&gt; [1] 2.755839</span></span>
<span id="cb14-4"><a href="#cb14-4"></a></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="co">## or use with P-value</span></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="kw"><a href="../reference/genomic_inflation.html">genomic_inflation</a></span>(<span class="dt">P =</span> giant<span class="op">$</span>P)</span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="co">#&gt; [1] 2.788867</span></span></code></pre></div>
</div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ul>
<li><p>Chang CC, Chow CC, Tellier LCAM, Vattikuti S, Purcell SM, Lee JJ (2015). <em>Second-generation PLINK: rising to the challenge of larger and richer datasets</em>. GigaScience, 4.</p></li>
<li><p>Naret O (2018). <em>CrowdAI / OpenSNP - height prediction challenge</em> [Data set]. Zenodo. <a href="http://doi.org/10.5281/zenodo.1442755" class="uri">http://doi.org/10.5281/zenodo.1442755</a></p></li>
<li><p>Yengo L, Sidorenko J, Kemper KE, Zheng Z, Wood AR, Weedon MN, Frayling TM, Hirschhorn J, Yang J, Visscher PM, the GIANT Consortium (2018). <em>Meta-analysis of genome-wide association studies for height and body mass index in ∼700000 individuals of European ancestry</em>. Human Molecular Genetics, Volume 27, Issue 20, 15 October 2018, Pages 3641–3649, <a href="https://doi.org/10.1093/hmg/ddy271" class="uri">https://doi.org/10.1093/hmg/ddy271</a>.</p></li>
<li><p>Gao X , Starmer J and Martin ER (2008). <em>A multiple testing correction method for genetic association studies using correlated single nucleotide polymorphisms</em>. Genet. Epidemiol., 32: 361-369, <a href="doi:10.1002/gepi.20310" class="uri">doi:10.1002/gepi.20310</a>.</p></li>
</ul>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Chromosome 1, 15’000’000 - 20’000’000 encompasses two regions with known human height associations: <a href="https://www.nature.com/articles/ng.3097#s2" class="uri">https://www.nature.com/articles/ng.3097#s2</a>.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#data">Data</a></li>
      <li><a href="#functions">Functions</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Sina Rüeger, Aurélien Macé, Zoltán Kutalik.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
